### PR TITLE
!cdd #17770 Use self recursive type for ReplicatedData

### PR DESCRIPTION
**Pull request by patriknw**

### CODE REVIEW

**Comment by patriknw** [comment]
This is only a prototype. Doesn't work yet.

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala [hunk 1]
```diff
@@ -164,8 +162,9 @@ final class ORMap[A <: ReplicatedData] private[akka] (
               s"[${thisValue.getClass.getName}], got [${thatValue.getClass.getName}]"
             throw new IllegalArgumentException(errMsg)
           }
+          val mergedValue = thisValue.merge(thatValue)
           // TODO can we get rid of these (safe) casts?
-          val mergedValue = thisValue.merge(thatValue.asInstanceOf[thisValue.T]).asInstanceOf[A]
+          //          val mergedValue = thisValue.merge(thatValue.asInstanceOf[thisValue.T]).asInstanceOf[A]
```
Comment (patriknw) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala]: This is the problem that I'm stuck on. Merging the values in ORMap. 

    type mismatch; found : thatValue.type (with underlying type A) required: _$6

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala [hunk 2]
```diff
@@ -33,12 +33,10 @@ object ORMap {
  * This class is immutable, i.e. "modifying" methods return a new instance.
  */
 @SerialVersionUID(1L)
-final class ORMap[A <: ReplicatedData] private[akka] (
+final class ORMap[A <: AbstractReplicatedData[_]] private[akka] (
```
Comment (viktorklang) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala]: Should that be `[A <: AbstractReplicatedData[A]` ?

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala [hunk 3]
```diff
@@ -33,12 +33,10 @@ object ORMap {
...
   private[akka] val keys: ORSet[String],
   private[akka] val values: Map[String, A])
-  extends ReplicatedData with ReplicatedDataSerialization with RemovedNodePruning {
-
-  type T = ORMap[A]
+  extends AbstractReplicatedData with ReplicatedDataSerialization { // FIXME with RemovedNodePruning {
```
Comment (viktorklang) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala]: No type param for AbstractReplicatedData?

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala (hunk 2 above)
Comment (ktoso) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala]: yeah, in other places as well, I'm helping out on getting this PR up to type-soundness right now :-)

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala (hunk 3 above)
Comment (patriknw) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala]: this might what caused the confusion in merge, thanks

**Comment by patriknw** [comment]
@ktoso It turns out that the internals in the Replicator just blows up with ugliness, existential types, and other things that I don't want there. Here is a good compromise. Using self recursive type only for the Java API, `AbstractReplicatedData`.

**Comment by patriknw** [comment]
This PR is not ready. I will also update the docs, and complete the TwoPhaseSet documentation example (serializer and stuff)

**Comment by ktoso** [comment]
I see, yeah you would have to carry around the ugly `A <: ...` everywhere I guess.
Makes sense to use it only for japi then.

**Comment by patriknw** [comment]
@ktoso I have completed the TwoPhaseSet documentation sample, which illustrates how to write a custom data type in java, including serialization. See bde5a06

**Comment by ktoso** [comment]
LGTM, awesome to see it worked out with the recursive type :)

**Comment by patriknw** [comment]
yeah, thanks for the great idea

**Comment by patriknw** [comment]
Refs #17770

**Pull request merged**